File: 30.substring-with-concatenation-of-all-words/solution.cpp
class Solution {
public:
	vector<int> findSubstring(string s, vector<string>& words) {
		vector<int> result;
		if (words.empty()) return result;
		multimap<string, int> cache;
		for (vector<string>::iterator it = words.begin(); it != words.end(); it++)
		{
			auto itr = cache.find(*it);
			if (itr == cache.end())
			{
				cache.emplace(*it, 1);
			} else
			{
				itr->second++;
			}
		}
		int count = 0;
		for (int i = 0; i + words.front().length() * words.size() <= s.length(); i++)
			if (check(s, i, cache, count, words.size())) result.push_back(i);
		return result;
	}
private:
	bool check(const string &s, int off, multimap<string, int> &cache, int count, const int &totalCount)
	{
		if (count == totalCount) return true;
		for (auto it = cache.begin(); it != cache.end(); it++)
		{
			if (it->second && startsWith(s, it->first, off))
			{
				it->second--;
				bool rtn = check(s, off + it->first.length(), cache, count + 1, totalCount);
				it->second++;
				if (rtn) return true;
			}
		}
		return false;
	}
	bool startsWith(const string &s, const string &sub, int off = 0)
	{
		if (s.length() - off < sub.length()) return false;
		for (int i = 0; i < sub.length(); i++)
			if (s[i+off] != sub[i]) return false;
		return true;
	}
};

File: 30.substring-with-concatenation-of-all-words/solution.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if len(s) == 0 or len(words) == 0:
            return []
        word_len = len(words[0])
        cache = {}
        for i in range(len(words)):
            if words[i] not in cache:
                cache[words[i]] = 1
            else:
                cache[words[i]] += 1
        result = []
        for i in range(len(s) - word_len * len(words) + 1):
            s0 = s[i:i + word_len]
            if s0 in cache and Solution.match(s[i:i + word_len * len(words)], cache.copy(), word_len):
                result.append(i)
        return result

    @staticmethod
    def match(s: str, cache: {str, int}, word_len: int) -> bool:
        for i in range(0, len(s), word_len):
            s0 = s[i:i + word_len]
            if s0 not in cache:
                return False
            cache[s0] -= 1
            if cache[s0] == 0:
                cache.pop(s0)
        return True
